examples/moses-pln-synergy/scm/README.md

```diff
@@ -76,16 +76,204 @@ And search for the following
    )
 ...
 
-scheme@(guile-user)> (cog-bind pln-rule-eliminate-neutral-element-hack)
-scheme@(guile-user)> (cog-bind pln-rule-eliminate-dangling-junctor-hack)
-scheme@(guile-user)> (cog-bind pln-rule-equivalence-hack)
-$6 = (SetLink
+scheme@(guile-user)> (cog-bind implication-lambda-distribution-rule)
+...
    (ImplicationLink (stv 1 0.99999982)
-      (PredicateNode "take-treatment-1" (stv 0.1 0.80000001))
-      (PredicateNode "take-compound-A" (stv 0.2 0.80000001))
+      (LambdaLink
+         (TypedVariableLink
+            (VariableNode "$X")
+            (TypeNode "ConceptNode")
+         )
+         (AndLink
+            (EvaluationLink
+               (PredicateNode "take")
+               (ListLink
+                  (VariableNode "$X")
+                  (ConceptNode "treatment-1")
+               )
+            )
+            (EvaluationLink (stv 1 0.99999982)
+               (PredicateNode "contain")
+               (ListLink
+                  (ConceptNode "treatment-1")
+                  (ConceptNode "compound-A")
+               )
+            )
+         )
+      )
+      (LambdaLink
+         (TypedVariableLink
+            (VariableNode "$X")
+            (TypeNode "ConceptNode")
+         )
```

Comment:
In predicate logic you can write
```predicate_logic
(ForAll x P(x)) -> (ForAll x Q(x))
```
it's possible, although usually people prefer to move all the quantifiers at the top level. Here we *cannot* do that, because
```atomese
ImplicationLink <TV>
   P
   Q
```
is not convertible to a single universal quantification. Instead it is a conditional quantification, the condition predicate is P, and the quantified predicate is Q. The formula is
```math
Sum_x P(x) and Q(x) / Sum_x P(x)
```
You cannot turn that into a universally quantification like in predicate logic, or even average quantification. So you cannot move all the quanfiers at the top level of the proposition.

Because of that you cannot do anything useful in PLN with free variables, cause, OK you've got a proposition with free variables (i.e. universally quantified), which means something that is not even useful to calculate the TVs of what matters to you, i.e. conditioned quantifications (i.e. ImplicationLinks).

So yes it make inference more complicated than with predicate logic, but I don't see another way, and I actually think it really is the simplest way.

By higher order I mean the ImplicationLink, not the domain, sorry for the confusion. The ImplicationLink works over sets, not individuals, just like map or fold, in this sense I mean higher order.